pymc3/distributions/multivariate.py

```diff
@@ -1911,39 +1869,29 @@ def _quaddist(self, value):
         else:
             onedim = False
 
-        delta = value - self.mu
-        if self._isEVD:
-            sqrt_quad = kron_dot(self.QTs, delta.T)
-            sqrt_quad = sqrt_quad / at.sqrt(self.eigs[:, None])
-            logdet = at.sum(at.log(self.eigs))
-        else:
-            sqrt_quad = kron_solve_lower(self.chols, delta.T)
-            logdet = 0
-            for chol_size, chol_diag in zip(self.sizes, self.chol_diags):
-                logchol = at.log(chol_diag) * self.N / chol_size
-                logdet += at.sum(2 * logchol)
+        delta = value - mu
+
+        eigh_iterable = map(eigh, covs)
+        eigs_sep, Qs = zip(*eigh_iterable)  # Unzip
+        Qs = list(map(at.as_tensor_variable, Qs))
+        QTs = list(map(at.transpose, Qs))
+
+        eigs_sep = list(map(at.as_tensor_variable, eigs_sep))
+        eigs = kron_diag(*eigs_sep)  # Combine separate eigs
+        eigs += sigma ** 2
+        N = eigs.shape[0]
+
+        sqrt_quad = kron_dot(QTs, delta.T)
+        sqrt_quad = sqrt_quad / at.sqrt(eigs[:, None])
+        logdet = at.sum(at.log(eigs))
+
         # Square each sample
         quad = at.batched_dot(sqrt_quad.T, sqrt_quad.T)
         if onedim:
             quad = quad[0]
-        return quad, logdet
 
-    def logp(self, value):
-        """
-        Calculate log-probability of Multivariate Normal distribution
-        with Kronecker-structured covariance at specified value.
-
-        Parameters
-        ----------
-        value: numeric
-            Value for which log-probability is calculated.
-
-        Returns
-        -------
-        TensorVariable
-        """
-        quad, logdet = self._quaddist(value)
-        return -(quad + logdet + self.N * at.log(2 * np.pi)) / 2.0
+        a = -(quad + logdet + N * at.log(2 * np.pi)) / 2.0
+        return at.as_tensor_variable(a)
```

Comment:
```suggestion
        return a
```
I don't think we need the `as_tensor_variable`